File: SwordDemoLibrary/app/src/main/java/com/example/swordlibrary/java/MainActivity2.java
package com.example.swordlibrary.java;

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;

import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import com.example.swordlibrary.R;
import com.sword.xlcwtest.EmulatorDetectUtils;
import com.sword.xlcwtest.EmulatorDetectUtilsUpdate1;
import com.sword.LogUtil;
import com.sword.xlcwtest.EmulatorDetectUtilsUpdate2;
import com.sword.xlcwtest.SystemUtil;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Locale;
import java.util.Random;
import java.util.UUID;


public class MainActivity2 extends AppCompatActivity {
    private static final String tag = "app-MainActivity";
    public static String CPUABI = null;

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        
        /*String t = "check emulator: " + EmulatorCheck.isEmulator(this);
        LogUtil.debug(tag, t);
        t += "\n\n" + EmulatorCheck.emulatorResult;
        ((TextView)findViewById(R.id.text)).setText(t);*/
        String debugInfo = "\n\n修改后 EmulatorDetectUtilsUpdate-1.isEmulator: " + EmulatorDetectUtilsUpdate1.isEmulator(this) 
            + "\n\n修改后 EmulatorDetectUtilsUpdate-2.isEmulator: " + EmulatorDetectUtilsUpdate2.isEmulator(this)
            + "\n\n修改前 EmulatorDetectUtils.isEmulator: " + EmulatorDetectUtils.isEmulator(this)
            + "\n\n自买量 SDK, System.getOs: " + SystemUtil.getOS();
        
        LogUtil.debug(debugInfo);
    }

    /**
     * 生成uuid
     * */
    public static String buildUUID() {
        String uuid = UUID.randomUUID().toString().replaceAll("-","");
        uuid = uuid + System.currentTimeMillis();
        return uuid;
    }

    /**
     * 生成uuid
     * */
    public static String buildUUID(String name) {
        String uuid = UUID.fromString(name).toString();
        uuid = uuid + System.currentTimeMillis();
        return uuid;
    }

    String s = "qwertyuiopasdfghjklzxcvbnmQWERRTYUIOPASDFGHJKLZXCVBNM1234567890";
    public String randomString(int count) {
        Random random = new Random();
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < count; i++) {
            result.append(s.charAt(random.nextInt(s.length())));
        }
        return result.toString();
    }

    private void requestPermission() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            Log.d("Sword", "申请权限");
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, 0);
        }
    }


    public static void getCPUABI() {
        if (CPUABI == null) {
            try {
                String os_cpuabi = new BufferedReader(new InputStreamReader(Runtime.getRuntime().exec("getprop ro.product.cpu.abi").getInputStream())).readLine();
                if (os_cpuabi.contains("x86")) {
                    CPUABI = "x86";
                } else if (os_cpuabi.contains("armeabi-v7a")) {
                    CPUABI = "armeabi-v7a";
                } else if (os_cpuabi.contains("arm64-v8a")){
                    CPUABI = "arm64-v8a";
                } else {
                    CPUABI = "armeabi";
                }
            } catch (Exception e) {
                CPUABI = "armeabi";
            }
            Log.d("sword", CPUABI);
        }
    }

    //打印 Local 信息
    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public String getLocal() {
        Locale locale = Locale.getDefault();
        return "language: " + locale.getLanguage() + "\nCountry: " + locale.getCountry() + "\n languageTag: " + locale.toLanguageTag() + "\ndisplayCountry: " + locale.getDisplayCountry() + "\ngetDisplayLanguage: " + locale.getDisplayLanguage() + "\ngetDisplayName: " + locale.getDisplayName() + "\ngetDisplayScript: " + locale.getDisplayScript() + "\ngetDisplayVariant: " + locale.getDisplayVariant() + "\ngetISO3Country: " + locale.getISO3Country() + "\ngetISO3Language: " + locale.getISO3Language() + "\ngetScript: " + locale.getScript() + "\ngetVariant: " + locale.getVariant();

    }
    
    //获取 Android 设备所在地区
    public void getIpAddress() {

        //1. 通过 NetworkInterface 获取 IP 地址，可能会获取不到
        //2. 通过 TelephonyManager 获取类似 MCC 或者 SIM 卡运营商所在地区的国家/地区代码
        //3. 通过 WifiManager 获取 IP 地址
        //4. 由服务器端下发 IP 地址，通过 IpSeekUtils 获取 IP 地址所在地区/国家
        //5. 访问外部连接
        //      http://pv.sohu.com/cityjson
        //      http://pv.sohu.com/cityjson?ie=utf-8
        //      http://ip.chinaz.com/getip.aspx


        Log.d("TAG", "shell 获取外网 ip：" + ShellAdbUtil.execShellCommand(false, "curl ifconfig.me").getSuccessString());
    }


}

File: SwordDemoLibrary/app/src/main/java/com/example/swordlibrary/MainActivity2.kt
package com.example.swordlibrary

import android.annotation.SuppressLint
import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.Button
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.LinearLayout.LayoutParams
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.children
import com.example.swordlibrary.view.*
import com.sword.LogUtil
import com.sword.initWindowSize
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.activity_main.view.*

class MainActivity2: AppCompatActivity() {
  private val tag = "MainActivity2"
  private var multiTouchView1: MultiTouchView1? = null
  private var multiTouchView2: MultiTouchView2? = null
  private var multiTouchView3: MultiTouchView3? = null
  private var multiTouchView4: MultiTouchView4? = null
  private var dragListenerGridView: DragListenerGridView? = null
  
  private var twoPager: TwoPager? = null
  
  private var contentContainer: FrameLayout? = null

  @SuppressLint("SetTextI18n", "InflateParams")
  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    setContentView(R.layout.activity_main)
    val rootView = findViewById<LinearLayout>(R.id.rootView)
    contentContainer = findViewById(R.id.contentView)
    rootView.addView(Button(this).apply {
      text = "MultiTouchView1 单点触控"
      setOnClickListener {
        contentView.visibility = View.VISIBLE
        if (multiTouchView1 == null) {
          multiTouchView1 = MultiTouchView1(
            this@MainActivity2)
          contentView.addView(multiTouchView1,
            FrameLayout.LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
          )
        }

        hideAllView()
        if (multiTouchView1?.visibility != View.VISIBLE) {
          multiTouchView1?.visibility = View.VISIBLE
        }
      } 
    }, LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT))
    
    rootView.addView(Button(this).apply {
      text = "MultiTouchView2 接力型滑动"
      setOnClickListener {
        contentView.visibility = View.VISIBLE
        if (multiTouchView2 == null) {
          multiTouchView2 = MultiTouchView2(
            this@MainActivity2)
          contentView.addView(multiTouchView2,
            FrameLayout.LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
          )
        }

        hideAllView()
        if (multiTouchView2?.visibility != View.VISIBLE) {
          multiTouchView2?.visibility = View.VISIBLE
        }
      }
    }, LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT))

    rootView.addView(Button(this).apply {
      text = "MultiTouchView3 协作型滑动"
      setOnClickListener {
        contentView.visibility = View.VISIBLE
        
        if (multiTouchView3 == null) {
          multiTouchView3 = MultiTouchView3(
            this@MainActivity2)
          contentView.addView(multiTouchView3,
            FrameLayout.LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
          )
        }

        hideAllView()
        if (multiTouchView3?.visibility != View.VISIBLE) {
          multiTouchView3?.visibility = View.VISIBLE
        }
      }
    }, LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT))
    
    rootView.addView(Button(this).apply {
      text = "MultiTouchView3 各自为战型"
      setOnClickListener {
        contentView.visibility = View.VISIBLE

        if (multiTouchView4 == null) {
          multiTouchView4 = MultiTouchView4(
            this@MainActivity2)
          contentView.addView(multiTouchView4,
            FrameLayout.LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
          )
        }

        hideAllView()
        if (multiTouchView4?.visibility != View.VISIBLE) {
          multiTouchView4?.visibility = View.VISIBLE
        }
      }
    }, LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT))

    rootView.addView(Button(this).apply {
      text = "TwoPager"
      setOnClickListener {
        contentView.visibility = View.VISIBLE

        if (twoPager == null) {
          twoPager = TwoPager(
            this@MainActivity2)
          
          twoPager?.addView(View(context).apply { 
            setBackgroundColor(Color.parseColor("#795548"))
          })
          twoPager?.addView(View(context).apply { 
            setBackgroundColor(Color.parseColor("#388E3C"))
          })
          contentView.addView(twoPager,
            FrameLayout.LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
          )
        }

        hideAllView()
        if (twoPager?.visibility != View.VISIBLE) {
          twoPager?.visibility = View.VISIBLE
        }
      }
    }, LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT))

    rootView.addView(Button(this).apply {
      text = "DragListenerGridView"
      setOnClickListener {
        contentView.visibility = View.VISIBLE

        if (dragListenerGridView == null) {
          dragListenerGridView = LayoutInflater.from(this@MainActivity2).inflate(R.layout.drag_listener_grid_view, null) as DragListenerGridView
          contentView.addView(
            dragListenerGridView, 
            LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))
        }

        LogUtil.debug("DragListenerGridView", "添加 DragListenerGridView")
        hideAllView()
        if (dragListenerGridView?.visibility != View.VISIBLE) {
          dragListenerGridView?.visibility = View.VISIBLE
        }
      }
    }, LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT))

    addDragHelperGridView()

    addDragToCollectLayout()

    addDragUpDownLayout()
    
    initWindowSize(this)
  }

  private var dragToCollectLayout: DragToCollectLayout? = null
  @SuppressLint("SetTextI18n")
  private fun addDragToCollectLayout() {
    rootView.addView(Button(this).apply {
      text = "DragToCollectLayout"
      isAllCaps = false
      setOnClickListener {
        contentContainer?.visibility = View.VISIBLE

        if (dragToCollectLayout == null) {
          dragToCollectLayout = LayoutInflater.from(this@MainActivity2).inflate(R.layout.drag_to_collect_layout, null) as DragToCollectLayout
          contentContainer?.addView(
            dragToCollectLayout,
            LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))
        }

        LogUtil.debug("DragHelperGridView", "添加 DragHelperGridView")
        hideAllView()
        if (dragToCollectLayout?.visibility != View.VISIBLE) {
          dragToCollectLayout?.visibility = View.VISIBLE
        }
      }
    }, LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT))
  }

  private var dragHelperGridView: DragHelperGridView? = null
  @SuppressLint("SetTextI18n", "InflateParams")
  private fun addDragHelperGridView() {
    rootView.addView(Button(this).apply {
      text = "DragHelperGridView"
      isAllCaps = false
      setOnClickListener {
        contentContainer?.visibility = View.VISIBLE

        if (dragHelperGridView == null) {
          dragHelperGridView = LayoutInflater.from(this@MainActivity2).inflate(R.layout.drag_helper_grid_view, null) as DragHelperGridView
          contentContainer?.addView(
            dragHelperGridView,
            LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))
        }

        LogUtil.debug("DragHelperGridView", "添加 DragHelperGridView")
        hideAllView()
        if (dragHelperGridView?.visibility != View.VISIBLE) {
          dragHelperGridView?.visibility = View.VISIBLE
        }
      }
    }, LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT))
  }

  private var dragUpDownLayout: DragUpDownLayout? = null
  @SuppressLint("SetTextI18n", "InflateParams")
  private fun addDragUpDownLayout() {
    rootView.addView(Button(this).apply {
      text = "DragUpDownLayout"
      isAllCaps = false
      setOnClickListener {
        contentContainer?.visibility = View.VISIBLE

        if (dragUpDownLayout == null) {
          dragUpDownLayout = LayoutInflater.from(this@MainActivity2).inflate(R.layout.drag_up_down, null) as DragUpDownLayout
          contentContainer?.addView(
            dragUpDownLayout,
            LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))
        }

        LogUtil.debug("DragHelperGridView", "添加 DragHelperGridView")
        hideAllView()
        if (dragUpDownLayout?.visibility != View.VISIBLE) {
          dragUpDownLayout?.visibility = View.VISIBLE
        }
      }
    }, LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT))
  }

  private fun hideAllView() {
    contentContainer?.children?.forEach { child -> 
      if (child.visibility != View.GONE) {
        child.visibility = View.GONE
      }
    }
  }

  override fun onBackPressed() {
    if (contentContainer?.visibility != View.GONE) {
      contentContainer?.visibility = View.GONE
    } else {
      super.onBackPressed()
    }
  }
}